### PR TITLE
Add action for anonymous user

### PR DESCRIPTION
**Pull request by dukenv0307**
### Details

We accept some actions for anonymous user
### Fixed Issues

$ https://github.com/Expensify/App/issues/21977
PROPOSAL: https://github.com/Expensify/App/issues/21977#issuecomment-1615006602

### Tests

1. Login with any account
2. Create a public room
3. Go to share code and copy the link of the room
4. Sign out and open the public room with the link above
5. Click on the header
6. Go to share code
7. Verify that the share code page appears instead of redirecting to sign in page
8. Hover any message and click on copy to clipboard
9. Verify that the user can copy this message to clipboard and not redirect to sign page
10. Hover on an attachment and click on download
11. Verify that the user can download successfully
- [x] Verify that no errors appear in the JS console

### Offline tests

Same as above
### QA Steps

1. Login with any account
2. Create a public room
3. Go to share code and copy the link of the room
4. Sign out and open the public room with the link above
5. Click on the header
6. Go to share code
7. Verify that the share code page appears instead of redirecting to sign in page
8. Hover any message and click on copy to clipboard
9. Verify that the user can copy this message to clipboard and not redirect to sign page
10. Hover on an attachment and click on download
11. Verify that the user can download successfully
- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

[Screencast from 12-07-2023 10:31:49.webm](https://github.com/Expensify/App/assets/129500732/ae1b3770-0e39-436e-99ed-8d492a046a8c)

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/129500732/9fd69b4f-c4f2-41e1-9628-6e2fcbf6620f

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/129500732/93275948-1e37-4f3e-98fd-691268440eb1

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/129500732/a2803ad8-8530-4158-ab37-fbb3b56778bd

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/129500732/eb688819-2749-4489-a5b8-fc9941240d36

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/129500732/82b6c7c3-529d-425d-9b22-897da11848c9

</details>

### CODE REVIEW

**Comment by Santhosh-Sellavel** [comment]
@dukenv0307 I think you don't understand.
What are the other pages that use `checkIfActionIsAllowed`, please gather that!

We don't need to make sure that the action can be performed by the anonymous user or not. And handle it where we can allow it.

Here is another similar duplicate [issue](https://github.com/Expensify/App/issues/22307).  It does not use the `checkIfActionIsAllowed`, a similar method is written in the file there.

Do you get it?

cc: @PauloGasparSv

**Comment by Santhosh-Sellavel** [comment]
Yes please investigate give us a list of such actions. We can decide and proceed

**Comment by Santhosh-Sellavel** [comment]
Also if you run into any similar issues please tag and ask them to hold for this.

**Comment by Santhosh-Sellavel** [comment]
Anywhere else a related check is used? @dukenv0307

**Comment by Santhosh-Sellavel** [comment]
> > Anywhere else a related check is used? @dukenv0307
> 
> The actions in FAB also use the check, but I think we don't need to take care of this because it must sign in to create action.

Cool just list all places we do the check so we don't miss anything, and let's wait for @PauloGasparSv and proceed with changes, thanks!

**Comment by PauloGasparSv** [comment]
> @PauloGasparSv Please review https://github.com/Expensify/App/pull/22321#issuecomment-1624066686 when you got a chance so we can moveforward, thanks!

Sure thing, I have to focus on a couple of other issues first so I may only have time to review this on Monday though

**Comment by PauloGasparSv** [comment]
Got some time to review this :D

> @Santhosh-Sellavel Here is the list of actions that an anonymous user can perform. Please help to check again if I miss some actions.

Are these the list of actions an anonymous user can perform or the actions where we check if they can perform by calling `checkIfActionIsAllowed`? I think the latter right?

> cc: @PauloGasparSv Let me know what is allowed and what is not, thanks!

@Santhosh-Sellavel @dukenv0307 I think we'll have to involve the team to decided that so before I tag someone can we please discuss this:

IMO the changes here are good enough and fix the original issue's problem. It also keeps all other logic restricted for anonymous users which seems to be the original intent of https://github.com/Expensify/App/issues/18318.

Shouldn't we just fix this specific problem? And then if someone reports a similar issue they can add that `isAnonymousAction` as `true` for whatever action was reported

I think we may not need to increase the scope of the change to all other actions, please share your thoughts!

**Comment by PauloGasparSv** [comment]
> @PauloGasparSv That makes sense to me. For now, I think we can fix this issue and fix it https://github.com/Expensify/App/issues/22307 too.

> @Santhosh-Sellavel We have another issue about anonymous action here https://github.com/Expensify/App/issues/22638.

Thks so much for catching those, I think @Santhosh-Sellavel already commented in one of them so I'll do the same for https://github.com/Expensify/App/issues/22638

Can you add a fix for those 2 here? (I don't think we fixed them already right?)

@Santhosh-Sellavel bump on [this](https://github.com/Expensify/App/pull/22321#issuecomment-1629442642) for fixing the original issue + https://github.com/Expensify/App/issues/22638 and https://github.com/Expensify/App/issues/22307 this time since these are the reported issues so far.

**Comment by Santhosh-Sellavel** [comment]
@PauloGasparSv @dukenv0307 Can we also include the option to copy the URL from the share code page as it makes more sense?

**Comment by Santhosh-Sellavel** [comment]
https://github.com/Expensify/App/assets/85645967/a0c2cdca-3ea7-4db4-b87a-627d6b18a37f

**Comment by PauloGasparSv** [comment]
> @PauloGasparSv @dukenv0307 Can we also include the option to copy the URL from the share code page as it makes more sense?

Yesss! I thought this was [fixed already since it was mentioned here](https://github.com/Expensify/App/pull/22321#issuecomment-1630903469) btw : )

**Comment by Santhosh-Sellavel** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/85645967/acf03a01-1e86-4047-aae4-476e47978532

</details>

<details>
<summary>Mobile Web - Chrome & Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/85645967/18a85114-34e3-4ce1-8b11-f783042c784d

</details>

<details>
<summary>Desktop</summary>

Couldn't test this, but that should not be blocking!

</details>

<details>
<summary>iOS & Android</summary>

https://github.com/Expensify/App/assets/85645967/1871f191-4ef5-4e64-94e9-dcf691378757

</details>

**Review by Santhosh-Sellavel — APPROVED** [review]
LGTM, thanks @dukenv0307!

All yours @PauloGasparSv!

**Review by PauloGasparSv — APPROVED** [review]
LGTM!

<details>
	Had a lot of problems testing in other platforms because of my setup here so skipped 
 most.

<details>

<summary>Web</summary>

https://github.com/Expensify/App/assets/6564265/c4569be2-ec88-4117-b0f9-8158f172408d


</details>

<details>

<summary>Mobile Web - Safari</summary>


https://github.com/Expensify/App/assets/6564265/f05158db-1c90-4afc-93b9-2f050c82c896



</details>

</details>

File: src/pages/home/report/ContextMenu/ContextMenuActions.js [hunk 1]
```diff
@@ -130,6 +133,7 @@ export default [
         getDescription: () => {},
     },
     {
+        isAnonymousAction: false,
```
Comment (sobitneupane) [inline src/pages/home/report/ContextMenu/ContextMenuActions.js:136]: In [this issue](https://github.com/Expensify/App/issues/27039), we decided to allow anonymous user to copy url and email.

**Pull request merged**